lib/cache.js

```diff
@@ -4,7 +4,25 @@ const log = require('npmlog')
 const pacote = require('pacote')
 const path = require('path')
 const rimraf = promisify(require('rimraf'))
+const semver = require('semver')
 const BaseCommand = require('./base-command.js')
+const npa = require('npm-package-arg')
+
+const searchCachePackage = (spec, cacheKeys) => {
+  const parsed = npa(spec)
+  if (parsed.rawSpec !== '' && parsed.type === 'tag')
+    throw new Error(`Cannot list cache keys for a tagged package.`)
+  const searchMFH = new RegExp(`^make-fetch-happen:request-cache:.*(?<!/[@a-zA-Z]+)/${parsed.name}/-/(.*)-(?<ver>.*).tgz$`)
```

Comment:
```javascript
> r
/\/-\/(.*)-(?<ver>.*).tgz$/
> '/-/foo-1.2.3-beta.tgz'.match(r)
[
  '/-/foo-1.2.3-beta.tgz',
  'foo-1.2.3',
  'beta',
  index: 0,
  input: '/-/foo-1.2.3-beta.tgz',
  groups: [Object: null prototype] { ver: 'beta' }
```

Reply:
The only reliable way to pull a version out of a tarball url is to know the name beforehand and explicitly strip out `name-` from the base filename

Reply:
I made a test that failed, implemented something similar to npm/registry-frontdoor PackageRouter.tarballHandler, and made sure the tests then passed.